// rc-form/lib/FormScope.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _createDOMForm = require('./createDOMForm');

var _createDOMForm2 = _interopRequireDefault(_createDOMForm);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var FormScope = function (_React$Component) {
  (0, _inherits3['default'])(FormScope, _React$Component);

  function FormScope() {
    (0, _classCallCheck3['default'])(this, FormScope);
    return (0, _possibleConstructorReturn3['default'])(this, (FormScope.__proto__ || Object.getPrototypeOf(FormScope)).apply(this, arguments));
  }

  (0, _createClass3['default'])(FormScope, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          children = _props.children,
          options = (0, _objectWithoutProperties3['default'])(_props, ['children']);

      var WrappedComponent = (0, _createDOMForm2['default'])(options)(function (_ref) {
        var form = _ref.form;

        (0, _extends3['default'])(_this2, form);
        return children(form);
      });
      return _react2['default'].createElement(WrappedComponent, null);
    }
  }]);
  return FormScope;
}(_react2['default'].Component);

FormScope.propTypes = {
  children: _propTypes2['default'].func,
  validateMessages: _propTypes2['default'].object,
  onFieldsChange: _propTypes2['default'].func,
  onValuesChange: _propTypes2['default'].func,
  mapProps: _propTypes2['default'].func,
  mapPropsToFields: _propTypes2['default'].func,
  fieldNameProp: _propTypes2['default'].string,
  fieldMetaProp: _propTypes2['default'].string,
  fieldDataProp: _propTypes2['default'].string,
  withRef: _propTypes2['default'].bool
};
exports['default'] = FormScope;
module.exports = exports['default'];
